### PR TITLE
Retry request once when meet exception.

### PR DESCRIPTION
**Pull request by Shylock-Hg**
Require https://github.com/vesoft-inc/nebula-common/pull/576

### CODE REVIEW

File: src/client/Session.cpp [hunk 1]
```diff
@@ -35,11 +61,10 @@ bool Session::ping() {
 }
 
 ErrorCode Session::retryConnect() {
-    pool_->giveBack(std::move(conn_));
-    conn_ = pool_->getConnection();
-    auto resp = conn_.authenticate(username_, password_);
-    sessionId_ = resp.sessionId != nullptr ? *resp.sessionId : -1;
-    return resp.errorCode;
+    if (!conn_.open()) {
```
Comment (laura-ding) [inline src/client/Session.cpp]: retry connect means to get an ok connection from the pool, no means to reopen the broken。

Comment (Shylock-Hg) [reply src/client/Session.cpp]: Fixed.

File: src/client/ConnectionPool.cpp [hunk 2]
```diff
@@ -30,19 +33,27 @@ void ConnectionPool::init(const std::vector<std::string> &addresses, const Confi
     }
     config_ = config;
     newConnection(0, config.maxConnectionPoolSize_);
+
+    keepAlive_.store(true, std::memory_order_release);
+    maintainer_ = std::thread(&ConnectionPool::keepAliveJob, this);
```
Comment (laura-ding) [inline src/client/ConnectionPool.cpp:38]: why use thread to add and remove the connection, it has a time delay.

File: src/client/Session.cpp [hunk 3]
```diff
@@ -34,12 +45,8 @@ bool Session::ping() {
     return conn_.ping();
 }
 
-ErrorCode Session::retryConnect() {
-    pool_->giveBack(std::move(conn_));
+void Session::retryConnect() {
     conn_ = pool_->getConnection();
```
Comment (laura-ding) [inline src/client/Session.cpp:49]: it maybe gets bad connection.

**Review by laura-ding — COMMENTED** [review]
add test for reconnect

File: src/client/ConnectionPool.cpp (hunk 2 above)
Comment (Shylock-Hg) [reply src/client/ConnectionPool.cpp:38]: It's used to reduce the time when retry connection, if don't keep alive of connection pool we need open new connection when retry connection.

File: src/client/Session.cpp (hunk 3 above)
Comment (Shylock-Hg) [reply src/client/Session.cpp:49]: If it's bad means we can't get available connection any way.

**Pull request closed without merging**